subsys/net/ip/icmp.c

```diff
@@ -0,0 +1,390 @@
+/** @file
+ * @brief ICMP related functions
+ */
+
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/* Use highest log level if both IPv4 and IPv6 are defined */
+#if defined(CONFIG_NET_IPV4)
+#define MY_LOG_LEVEL CONFIG_NET_ICMPV4_LOG_LEVEL
+#endif
+
+#if defined(CONFIG_NET_IPV6)
+#if defined(MY_LOG_LEVEL)
+#if MY_LOG_LEVEL < CONFIG_NET_ICMPV6_LOG_LEVEL
+#define MY_LOG_LEVEL CONFIG_NET_ICMPV6_LOG_LEVEL
+#endif
+#else
+#define MY_LOG_LEVEL CONFIG_NET_ICMPV6_LOG_LEVEL
+#endif
+#endif
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(net_icmp, MY_LOG_LEVEL);
+
+#include <errno.h>
+#include <zephyr/sys/slist.h>
+#include <zephyr/net/net_pkt.h>
+#include <zephyr/net/icmp.h>
+
+#include "net_private.h"
+#include "icmpv6.h"
+#include "icmpv4.h"
+#include "ipv4.h"
+#include "ipv6.h"
+#include "net_stats.h"
+
+static K_MUTEX_DEFINE(lock);
+static sys_slist_t handlers;
+
+#define PKT_WAIT_TIME K_SECONDS(1)
+
+int net_icmp_init_ctx(struct net_icmp_ctx *ctx, uint8_t type, uint8_t code,
+		      net_icmp_handler_t handler)
+{
+	if (ctx == NULL || handler == NULL) {
+		return -EINVAL;
+	}
+
+	if (ctx->handler != NULL) {
+		return -EALREADY;
+	}
+
+	memset(ctx, 0, sizeof(struct net_icmp_ctx));
+
+	ctx->handler = handler;
+	ctx->type = type;
+	ctx->code = code;
+
+	k_mutex_lock(&lock, K_FOREVER);
+
+	sys_slist_prepend(&handlers, &ctx->node);
+
+	k_mutex_unlock(&lock);
+
+	return 0;
+}
+
+int net_icmp_cleanup_ctx(struct net_icmp_ctx *ctx)
+{
+	if (ctx == NULL) {
+		return -EINVAL;
+	}
+
+	k_mutex_lock(&lock, K_FOREVER);
+
+	sys_slist_find_and_remove(&handlers, &ctx->node);
+
+	k_mutex_unlock(&lock);
+
+	memset(ctx, 0, sizeof(struct net_icmp_ctx));
+
+	return 0;
+}
+
+static int send_icmpv4_echo_request(struct net_icmp_ctx *ctx,
+				    struct net_if *iface,
+				    struct in_addr *dst,
+				    uint16_t identifier,
+				    uint16_t sequence,
+				    uint8_t tos,
+				    int priority,
+				    const void *data,
+				    size_t data_size,
+				    void *user_data)
+{
+	NET_PKT_DATA_ACCESS_CONTIGUOUS_DEFINE(icmpv4_access,
+					      struct net_icmpv4_echo_req);
+	int ret = -ENOBUFS;
+	struct net_icmpv4_echo_req *echo_req;
+	const struct in_addr *src;
+	struct net_pkt *pkt;
+
+	if (IS_ENABLED(CONFIG_NET_OFFLOAD) && net_if_is_ip_offloaded(iface)) {
+		/* XXX: fixme so that we can send offloaded messages too */
+		return -ENOTSUP;
+	}
+
+	if (!iface->config.ip.ipv4) {
+		return -ENETUNREACH;
+	}
+
+	/* Take the first address of the network interface */
+	src = &iface->config.ip.ipv4->unicast[0].address.in_addr;
```

Comment:
I see it's how it is implemented right now, but wouldn't it be better to use `net_if_ipv4_select_src_addr()` here?

Reply:
Yes, will fix this and the other issues reported by twister in next version. I have also proposal for the offloaded API in next version. I will also prepare test cases for this.